[App Extensibility] Detect conflicting homepage url (@W-17599554@)

vmarta opened the pull request:
This PR handles the scenario of an app extension adding a homepage at the root `/`. For example: the typescript-minimal template has its own homepage, but if you use the chakra-storefront extension, it'll also try to add its own homepage at the same url. Thus, there are 2 routes that target the same `/` path.

With this PR, there will no longer be a conflict in the routes:
- When the dev server runs, it'll open at `/__pwa-kit/start` path by default. The base template's homepage will thus be rendered.
- while the storefront extension's homepage is still at the `/` route.

- [x] Fix tests and maybe add new ones

## How to test-drive the PR

1. Generate a project with `node packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js --outputDir /tmp/foo`
2. Select chakra-storefront as one of the extensions to install
3. Then run your generated project via `npm start`
4. Verify that the resulting page is the typescript-minimal page.

With storefront extension installed:
![Arc 2025-01-24 at 13 27 27](https://github.com/user-attachments/assets/8e3ada18-0647-429e-8ce6-6df9e6f800c1)

Without any extensions installed:
![Arc 2025-01-24 at 13 26 53](https://github.com/user-attachments/assets/c40d016f-b9ac-4d99-b526-9cfc2d76fa32)

vmarta committed:
Storefront extension: detects an existing homepage at same url

If there is, remove its route.

vmarta committed:
Only splice if the original home is found

vmarta commented on packages/extension-chakra-storefront/src/setup-app.tsx:
```diff
@@ -112,7 +112,17 @@ class ChakraStorefront extends ApplicationExtension<Config> {
         const config = this.getConfig()
 
-        return configureRoutes(allRoutes, config, {
+        const indexOfOriginalHome = allRoutes.findIndex(
+            (route) =>
+                route.path === (config.pages.Home && config.pages.Home.path) &&
+                route.component !== Pages.Home
+        )
+        const routes = allRoutes.slice()
+        if (indexOfOriginalHome >= 0) {
+            routes.splice(indexOfOriginalHome, 1)
+        }
+
+        return configureRoutes(routes, config, {
```
A dumb simple way that uses the existing mechanism `beforeRouteMatch` to update the routes and delete the original home route.

The main question that came to my mind is: whose responsibility is it to detect whether there's a conflicting homepage route? I think it should be the app extension that is responsible for that.

bendvc replied:
I'm not sure this is the right approach. There is an established pattern for priority when it comes to route matching and extensions, e.g. the most recent extension added to the extensions list has the highest priority for using a particular route, and the base project is ultimately the highest priority. 

There is nothing wrong per-say with our current scenario, it just so happens that that we have 2 projects using the same path and one of them is going to win. Thats just how it works. I don't think we should be looking at ways for the extension to solve this problem, instead we will eventually have logging that will inform the end user that there either duplicate routes and inform then with one resolved or won, and provide then with information to change it if that isn't the desired behaviour. 

But what we have here is that our base project shouldn't have what I'll call a public route because there will always be a change that it will conflict with a route provided by an extension that we don't know about. 

Thats why the outline solution I shared previously is essentially to make sure that the base template will never conflict it's home route because it's "private" in that we are making it a hard to guess name like `__getting-started`. 

But we would have to make sure that this page isn't always navigated to every time we open the dev server as what isn't the desired behaviour, especially if an extension is registering the `/` path. So we want to make it a one time use thing. 

In general we are solving this with a feature that can be used for other things also. The solution you are proposing here will only work for our current situation, but if there is another extension that we add that registers the `/` path, then the code wouldn't work as it's only replacing one of those occurrences.

bendvc replied:
Additionally when we look at the flow that is going to occur in the real world. Developers are going to generate projects and then once they start creating pages and whatnot, they will most likely delete the home page that is generated. That would mean that you have code in the extension that is never really going to do anything.

vmarta committed:
Load the index route if exists. Otherwise load the start page from base template

vmarta committed:
Revert changes

vmarta committed:
Bonus: fix invalid html and also react hydration error

The <ul> tag cannot be nested inside a <p>.

vmarta committed:
typescript-minimal homepage now renders properly inside storefront

We needed to isolate the css styles to prevent leaking.

vmarta committed:
Allow users to configure the starting path for the dev server

vmarta committed:
Update log message

vmarta committed:
A more minimal homepage

- Removed the gradient background
- Content is now mostly static, but dynamically list the installed app extensions

vmarta committed:
Add button to link to the docs

vmarta committed:
Remove todo

vmarta commented on packages/template-typescript-minimal/app/pages/home.tsx:
```diff
@@ -4,171 +4,96 @@
+                            </a>
+                        </p>
+
+                        <p>You have installed the following application extensions:</p>
```
The typescript-minimal homepage now has a more minimal content.

vmarta commented on packages/template-typescript-minimal/app/pages/home.tsx:
```diff
@@ -4,171 +4,96 @@
-  100% { opacity: 1 }
-}
-.fade-in {
+.content {
```
Fixed and constrained the css selectors to avoid the styles leaking to other unexpected elements.

vmarta commented on packages/template-typescript-minimal/package.json:
```diff
@@ -48,6 +49,166 @@
   "mobify": {
     "app": {
       "extensions": [
+        ["@salesforce/extension-chakra-storefront", {
```
I'll revert this change before the PR is merged.

bendvc commented on packages/template-typescript-minimal/app/pages/home.tsx:
```diff
@@ -4,171 +4,96 @@
     margin-left: 5em;
     margin-right: 3em;
 }
+.button {
+    display: inline-block;
+    padding: 10px 20px;
+    font-size: 16px;
+    font-weight: bold;
+    color: white;
+    background-color: #0176D3;
+    text-align: center;
+    text-decoration: none;
+    border-radius: 5px;
+}
```
I like that it looks like a chakra button without being one 👍

bendvc commented on packages/template-typescript-minimal/app/pages/home.tsx:
```diff
@@ -4,171 +4,96 @@
-                    </p>
+                                ))
+                            ) : (
+                                <li>None</li>
```
This "none" is very computer sounding. Maybe we can make that look a little better. Like:

If you have extensions installed:

```
You have 2 extensions installed.
    - Chakra Storefront
    - Chakra Store Locator
```
If you don't have extensions installed:

```
You currently do not have any extensions installed.
```

bendvc replied:
This by no means is what we'll have in the end. But it's a little cleaner. It would be nice to have friendly names like my examples above, but using the package name is ok for now.

vmarta committed:
Fix linting error

It was complaining about the unnecessary type assertion.

bendvc commented on packages/template-typescript-minimal/app/pages/home.tsx:
```diff
@@ -4,171 +4,96 @@
 
-const Home = ({value}: Props) => {
-    const [counter, setCounter] = useState(0)
+const Home = () => {
```
Is there any reason why we don't want to called this "Start" or "GettingStarted"? I personally like the later, but I would that the path is in alignment with the page name.. so /_pwa-kit/getting-started and GettingStarted or /_pwa-kit/start and Start

alexvuong replied:
Agreed with Ben on this

alexvuong commented on packages/template-typescript-minimal/app/ssr.js:
```diff
@@ -27,6 +27,9 @@ const options = {
     protocol: 'http',
 
+    // The path that the local dev server would open initially
+    startPath: '/__pwa-kit/start',
```
Nit: it seems like there is a small chance the Home page will result in a 404 if the value here is mismatched with the one in the path in `routes`. I don't think creating a constant would make much sense since it is likely to change when devepers starts to alternate the routes during developerment. Maybe leaving a comment in the routes to give more context for this value?

vmarta replied:
Good point. I've added a comment in the routes file to clarify what `__pwa-kit/getting-started` path is really for.

vmarta committed:
Fix CI test

vmarta committed:
Tweak messaging

vmarta committed:
Rename Home to GettingStarted

vmarta committed:
Add comment to clarify

vmarta committed:
Make sure to encode the 'path' query param

vmarta committed:
Update template name

vmarta committed:
Set '/' as default value for startPath

vmarta committed:
Back to no extensions installed

vmarta committed:
Back to no extensions installed

vmarta committed:
Bonus: update config files for Store Locator

vmarta commented on packages/extension-chakra-store-locator/config/sample.json:
```diff
@@ -16,5 +16,16 @@
         }
-    ]
+    ],
+    "commerceApi": {
+        "proxyPath": "/mobify/proxy/api",
+        "parameters": {
+            "shortCode": "8o7m175y",
+            "clientId": "c9c45bfd-0ed3-4aa2-9971-40f88962b836",
+            "organizationId": "f_ecom_zzrf_001",
+            "siteId": "RefArchGlobal",
+            "locale": "en-GB",
+            "currency": "USD"
+        }
+    }
```
Since I reset the extensions config found in typescript-minimal's package.json, I brought it over to here where it belongs.
